### PR TITLE
KEYCLOAK-18845 Remove key type in map storage

### PR DESCRIPTION
**Pull request by hmlnarik**
This PR is to simplify `MapStorage` interface by removing the key type from the generics as it should be an implementation detail. Specifically see [model/map/src/main/java/org/keycloak/models/map/storage/MapStorage.java](https://github.com/keycloak/keycloak/pull/8288/files#diff-91ca21d5c074b826e13ccf8efc1fd50ec91d2ce30c40a77ff463cd2e16d76f7b)

### CODE REVIEW

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapKeycloakTransaction.java [hunk 1]
```diff
@@ -207,16 +209,21 @@ private V getUpdated(V orig) {
 
     @Override
     public V create(V value) {
-        K key = value.getId();
+        String key = value.getId();
+        if (key == null) {
+            K newKey = keyConvertor.yieldNewUniqueKey();
+            key = keyConvertor.keyToString(newKey);
+            value = Serialization.from(value, key);
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapKeycloakTransaction.java:216]: Can this be solved by introducing `void setId(String sKey)` to AbstractEntity?

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java [hunk 2]
```diff
@@ -118,26 +113,26 @@ private ModelException userDoesntExistException() {
...
     }
 
-    private Optional<MapUserEntity<K>> getEntityById(RealmModel realm, K id) {
-        MapUserEntity<K> mapUserEntity = tx.read(id);
-        if (mapUserEntity != null && entityRealmFilter(realm).test(mapUserEntity)) {
-            return Optional.of(mapUserEntity);
-        }
-
-        return Optional.empty();
+    private Optional<MapUserEntity> getRegisteredEntityById(RealmModel realm, String id) {
```
Comment (mhajas) [inline model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:134]: Wrong rebase? this method should be removed.

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you very much for the PR @hmlnarik. The changes look great, the readability Map*Providers is much then before. I added a few comments/questions into the code.

File: model/map/src/main/java/org/keycloak/models/map/authSession/MapRootAuthenticationSessionProviderFactory.java [hunk 3]
```diff
@@ -26,7 +26,7 @@
 /**
  * @author <a href="mailto:mkanis@redhat.com">Martin Kanis</a>
  */
-public class MapRootAuthenticationSessionProviderFactory<K> extends AbstractMapProviderFactory<AuthenticationSessionProvider, K, MapRootAuthenticationSessionEntity<K>, RootAuthenticationSessionModel>
+public class MapRootAuthenticationSessionProviderFactory<K> extends AbstractMapProviderFactory<AuthenticationSessionProvider, K, MapRootAuthenticationSessionEntity, RootAuthenticationSessionModel>
```
[the remark this answers is not in the document]
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/authSession/MapRootAuthenticationSessionProviderFactory.java:29]: good catch

File: model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java (hunk 2 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/user/MapUserProvider.java:134]: good catch!

File: model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapKeycloakTransaction.java (hunk 1 above)
Comment (hmlnarik) [reply model/map/src/main/java/org/keycloak/models/map/storage/chm/ConcurrentHashMapKeycloakTransaction.java:216]: FTR: In here we need always to make a copy of a value from the parameter, to ensure that the caller does not get the direct reference to the entered entity (which could be of a completely different type than `Map*EntityImpl`)

**Review by mhajas — APPROVED** [review]
Thanks for the changes @hmlnarik!

**Pull request merged**